src/proto/h2/bdp.rs

```diff
@@ -0,0 +1,190 @@
+            bdp: initial_window,
+            max_bandwidth: 0.0,
+            shared,
+            samples: 0,
+            rtt: 0.0,
+        },
+    )
+}
+
+#[derive(Clone)]
+pub(crate) struct Sampler {
+    shared: Weak<Mutex<Shared>>,
+}
+
+pub(super) struct Estimator {
+    shared: Arc<Mutex<Shared>>,
+
+    /// Current BDP in bytes
+    bdp: u32,
+    /// Largest bandwidth we've seen so far.
+    max_bandwidth: f64,
+    /// Count of samples made (ping sent and received)
+    samples: usize,
+    /// Round trip time in seconds
+    rtt: f64,
+}
+
+struct Shared {
+    bytes: usize,
+    ping_pong: PingPong,
+    ping_sent: bool,
+    sent_at: Instant,
+}
+
+impl Sampler {
+    pub(crate) fn sample(&self, bytes: usize) {
+        let shared = if let Some(shared) = self.shared.upgrade() {
+            shared
+        } else {
+            return;
+        };
+
+        let mut inner = shared.lock().unwrap();
+
+        if !inner.ping_sent {
+            if let Ok(()) = inner.ping_pong.send_ping(Ping::opaque()) {
+                inner.ping_sent = true;
+                inner.sent_at = Instant::now();
+                trace!("sending BDP ping");
+            } else {
+                return;
+            }
+        }
+
+        inner.bytes += bytes;
+    }
+
+    pub(crate) fn is_enabled(&self) -> bool {
+        self.shared.strong_count() > 0
+    }
+}
+
+impl Estimator {
+    pub(super) fn poll_estimate(&mut self, cx: &mut task::Context<'_>) -> Poll<WindowSize> {
+        let mut inner = self.shared.lock().unwrap();
+        if !inner.ping_sent {
+            // XXX: this doesn't register a waker...?
+            return Poll::Pending;
+        }
+
+        let (bytes, rtt) = match ready!(inner.ping_pong.poll_pong(cx)) {
+            Ok(_pong) => {
+                let rtt = inner.sent_at.elapsed();
+                let bytes = inner.bytes;
+                inner.bytes = 0;
+                inner.ping_sent = false;
+                self.samples += 1;
+                trace!("received BDP ack; bytes = {}, rtt = {:?}", bytes, rtt);
+                (bytes, rtt)
+            }
+            Err(e) => {
+                debug!("bdp pong error: {}", e);
+                return Poll::Pending;
+            }
+        };
+
+        drop(inner);
+
+        if let Some(bdp) = self.calculate(bytes, rtt) {
+            Poll::Ready(bdp)
+        } else {
+            // XXX: this doesn't register a waker...?
+            Poll::Pending
+        }
+    }
+
+    fn calculate(&mut self, bytes: usize, rtt: Duration) -> Option<WindowSize> {
+        // No need to do any math if we're at the limit.
+        if self.bdp as usize == BDP_LIMIT {
+            return None;
+        }
+
+        // average the rtt
+        let rtt = seconds(rtt);
+        if self.samples < 10 {
+            // Average the first 10 samples
+            self.rtt += (rtt - self.rtt) / (self.samples as f64);
+        } else {
+            self.rtt += (rtt - self.rtt) / 0.9;
```

Comment:
Is this trying to be an exponential moving average? Should this be `* 0.9` then?

Reply:
Yea, I inversed it by accident. It's fixed in the latest release.

Reply:
Ah my bad, I didn't notice because I was using an old release.